promql/query_logger.go

```diff
@@ -184,3 +190,38 @@ func (tracker ActiveQueryTracker) Insert(query string) int {
 
 	return i
 }
+
+// reloadQueryLogger returns a new Query Logger and a file to be closed once the
+// query loggers have been swapped.
+func reloadQueryLogger(ql *QueryLogger, s string) (*QueryLogger, *os.File, error) {
+	var (
+		timestampFormat = log.TimestampFormat(
+			func() time.Time { return time.Now().UTC() },
+			"2006-01-02T15:04:05.000Z07:00",
+		)
+		l        = &QueryLogger{}
+		prevFile *os.File
+	)
+
+	if ql != nil {
+		prevFile = ql.file
+	}
+
+	if s == "" {
+		return nil, prevFile, nil
+	}
+
+	// We always open the file again; so that a Prometheus reload can be
+	// called by external programs when they rotate the query log file.
+	f, err := os.OpenFile(s, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
```

Comment:
Don't make presumptions about permissions, always 0o666. It's umask's responsibility to limit this.